internal/restic/backend.go

```diff
@@ -20,8 +20,11 @@ type Backend interface {
 	// Close the backend
 	Close() error
 
-	// Save stores the data in the backend under the given handle.
-	Save(ctx context.Context, h Handle, rd io.Reader) error
+	// Save stores the data returned by the reader returned from fn in the
+	// backend under the given handle. The function fn may be called several
+	// times to retry the operation in case an error occurs. When fn returns an
+	// error, the process is aborted and the error is returned to the caller.
+	Save(ctx context.Context, h Handle, length int, fn func() (io.Reader, error)) error
```

Comment:
The proposed API does not look right, to be honest. If `fn` provides the content to save, it should provide both the content reader and the content length (and the error).

I also think adding `length` parameter is orthogonal to `fn` callback.

Reply:
Hm. But for the backend, we **always** know beforehand how much data we're going to save, and that's not going to change on the third try. `fn` is just a way to "reset" the reader, so that is left out of the backend code.

Reply:
I am nitpicking, but `length` parameter and the reader returned by `fn` represent related features of the content being saved. I think the API will be easier to understand and use correctly if both of these features were accessible the same way.

Personally, I'd introduce `ContentReader` interface with `Length()`, `Read(...)` and `Reset()` methods and define save as:

```go
Save(ctx context.Context, h Handle, content ContentReader) error
```

I think this will remove all the ugliness you want to remove, without the extra noise required to use`fn` parameter.

Reply:
Uh, that is a nice idea! Thank you very much! I'll try it later